Add Hebrew Support

tnsilver opened the pull request:
### Reference Issues / PR's

* adds Hebrew language support
* suggested promotion of minor version to  __1.8.1__

### What does this implement/fix
 
* adds package `pl.allegro.finance.tradukisto.internal.languages.hebrew`;
* adds class `pl.allegro.finance.tradukisto.internal.languages.hebrew.HebrewBigDecimalToBankingMoneyConverter`
* adds class `pl.allegro.finance.tradukisto.internal.languages.hebrew.HebrewHundredsToWordsConverter`
* adds class `pl.allegro.finance.tradukisto.internal.languages.hebrew.HebrewIntegerToWordsConverter`
* adds class `pl.allegro.finance.tradukisto.internal.languages.hebrew.HebrewValues`
* modifies enum `pl.allegro.finance.tradukisto.MoneyConverters` - adds member `HEBREW_BANKING_MONEY_VALUE`
* modifies enum `pl.allegro.finance.tradukisto.ValueConverters` - adds member `HEBREW_INTEGER`
* modifies class `pl.allegro.finance.tradukisto.internal.Container` - adds method `hebrewContainer()`

### What does this test?

* adds package `pl.allegro.finance.tradukisto.internal.languages.hebrew`
* adds class  `pl.allegro.finance.tradukisto.internal.languages.hebrew.HebrewBigDecimalToBankingMoneyConverterTest`
* adds class `pl.allegro.finance.tradukisto.internal.languages.hebrew.HebrewIntegerToStringConverterTest`

### Additional Comments

* adds the line `apply plugin: "eclipse"` to __build.gradle__ for eclipse plugin support
* modifies `.gitignore` to ignore eclipse files `.project` and `.classpath` and directory  `.settings`

tnsilver commented:
### Reference Issues / PR's

* fixes jacoco coverage
* adds tests in `pl.allegro.finance.tradukisto.internal.languages.hebrew.HebrewBigDecimalToBankingMoneyConverterTest`
* adds tests in `pl.allegro.finance.tradukisto.internal.languages.hebrew.HebrewIntegerToStringConverterTest`

wjur commented on build.gradle:
```diff
@@ -20,6 +20,7 @@ plugins {
 }
 
 apply plugin: 'io.codearte.nexus-staging'
+apply plugin: "eclipse"
```
Please use single quote to be consistent with the rest of file.

tnsilver replied:
done

wjur commented on src/main/java/pl/allegro/finance/tradukisto/internal/languages/hebrew/HebrewHundredsToWordsConverter.java:
```diff
@@ -0,0 +1,90 @@
+
+    @Override
+    public String asWords(Integer value, GenderType genderType) {
+        // System.out.printf("HebrewHundredsToWordsConverter.asWords() %s,%d%n",
```
Please delete comments.

tnsilver replied:
removed comments

tnsilver commented on CHANGELOG.md:
```diff
@@ -95,7 +95,7 @@ Fixed:
 ## 1.7.0 (2019-02-04)
 
-Added support for:
+Added support for:19
```
Thanks for the comments. I will fix nearly all of them on this week end.

I need you to understand that Hebrew is a very complex language in terms of gender. Single digits in general are addressed in female form - but if they address a masculine object, the form take male gender. Hundreds are female again, 
so the number 801 is read "eight (female) hundreds and one (female)" - but thousands are male and so are millions and billions, so 7801 is read "seven (male) eight (female) and one (female)". Also the "and" may appear more than once, in specific places. It's a nightmare language!

Currency is even more complex because the currency symbol for ILS is considered male and this changes the gender form, yet the cents (1/100) form of the currency is considered female and it changes the gender form once again, so the BigDecimalToBankingMoneyConverter is absolutely required.

The protected methods had to be there for the sole purpose of passing the Travis coverage requirements. Without them, I can't test "throws" branches and I don't pass the 100% coverage. This is really weird! 100% coverage is, by my standard, super strict. We usually settle for 76%. It's nearly impossible to test all "if" and "throws" branches and when Travis forces me to do so - I need methods to allow me to.

Kind regards, Tom Silverman

tnsilver commented on src/main/java/pl/allegro/finance/tradukisto/ValueConverters.java:
```diff
@@ -21,19 +22,13 @@
-    SERBIAN_INTEGER(serbianContainer().getNumbersConverter()),
-    FRENCH_INTEGER(frenchContainer().getNumbersConverter()),
-    TURKISH_INTEGER(turkishContainer().getNumbersConverter());
+    ENGLISH_INTEGER(englishContainer().getNumbersConverter()), GERMAN_INTEGER(germanContainer().getNumbersConverter()),
```
reformatted

tnsilver commented on src/test/groovy/pl/allegro/finance/tradukisto/internal/languages/hebrew/HebrewIntegerToStringConverterTest.groovy:
```diff
@@ -0,0 +1,247 @@
+import spock.lang.Unroll
+
+/**
+ * The class HebrewIntegerToStringConverterTest is a test case for the functionality of the HebrewIntegerToStringConverter
```
all comments removed

tnsilver commented on src/main/java/pl/allegro/finance/tradukisto/internal/languages/hebrew/HebrewIntegerToWordsConverter.java:
```diff
@@ -0,0 +1,179 @@
+        return other == 0;
+    }
+
+    protected HebrewValues hebrewValues() {
```
agreed - but required for for coverage tests

tnsilver commented on src/main/java/pl/allegro/finance/tradukisto/internal/languages/hebrew/HebrewIntegerToWordsConverter.java:
```diff
@@ -0,0 +1,179 @@
+        if (Range.closed(1000000000, Integer.MAX_VALUE).contains(value)) {
+            return billionsAsString(value);
+        }
+        // System.out.printf("HebrewIntegerToWordsConverter.asWords() SUPER value:
```
removed

tnsilver commented on src/main/java/pl/allegro/finance/tradukisto/internal/Container.java:
```diff
@@ -55,13 +59,11 @@ public static Container czechContainer() {
-        BigDecimalToStringConverter bigDecimalBankingMoneyValueConverter = new BigDecimalToBankingMoneyConverter(
-                integerConverter,
-                czechValues.currency());
+        IntegerToStringConverter integerConverter = new CzechIntegerToWordsConverter(containerForBigNumbers.getNumbersConverter(),
```
done

tnsilver commented on src/main/java/pl/allegro/finance/tradukisto/internal/Container.java:
```diff
@@ -87,48 +90,71 @@ public static Container frenchContainer() {
+
+        HebrewHundredsToWordsConverter hundredsToWordsConverter = new HebrewHundredsToWordsConverter(conversionValues);
+        HebrewIntegerToWordsConverter intToWordsConverter = new HebrewIntegerToWordsConverter(hundredsToWordsConverter,
+            conversionValues);
```
done

tnsilver commented on src/main/java/pl/allegro/finance/tradukisto/internal/Container.java:
```diff
@@ -87,48 +90,71 @@ public static Container frenchContainer() {
     }
 
+    /**
+     * adds Hebrew language support
```
I don't know... The whole disregard for documentation and the relying on naming conventions is beyond me. I let go of developers who do not document their code and it's made clear to them that JavaDoc is part of their job same as coding. Further more, neither me, my colleagues, company or any that I've worked for accept undocumented code. Then again... It's your conventions. Be my guest. Removed.

tnsilver commented on src/main/java/pl/allegro/finance/tradukisto/internal/Container.java:
```diff
@@ -87,48 +90,71 @@ public static Container frenchContainer() {
+    /**
+     * adds Hebrew language support
+     *
+     * @author <a href="mailto:tnsilver@gmail.com">T.N.Silverman</a>
```
removed

tnsilver commented on src/main/java/pl/allegro/finance/tradukisto/internal/Container.java:
```diff
@@ -87,48 +90,71 @@ public static Container frenchContainer() {
+     * adds Hebrew language support
+     *
+     * @author <a href="mailto:tnsilver@gmail.com">T.N.Silverman</a>
+     * @since 1.8.1
```
done

tnsilver commented on src/main/java/pl/allegro/finance/tradukisto/internal/languages/hebrew/HebrewBigDecimalToBankingMoneyConverter.java:
```diff
@@ -0,0 +1,83 @@
+
+    @Override
+    public String asWords(BigDecimal value) {
+        BigDecimal scaled = value.setScale(2, BigDecimal.ROUND_DOWN);
```
added validation for 2 decimal places as other languages

deejay1 commented:
The bin directory probably shouldn't be included in the PR.

tnsilver commented:
Really? That's what you have to tell me months after? Take the dang bin
directory out if you want to. Do whatever you want with it. Integrate the
pull with the master branch or not. Your choice. I have done my share. It's
time you guys move your rear end too.

Regards. Tom.

On Sat, Oct 17, 2020 at 10:00 PM Łukasz Jernaś ***@***.***>
wrote:

> The bin directory probably shouldn't be included in the PR.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/allegro/tradukisto/pull/38#issuecomment-711065393>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAB2QHMRRVHXJ2MK4PEXGXLSLHSVDANCNFSM4PYQ6XGA>
> .
>

wjur commented:
@tnsilver, I understand your frustration. You contributed to an open source library and after more than a year, didn’t get to master, but got some silly, meaningless approval. You haven't been deleting the `bin` directory for awhile, but I don’t have resources to clean up the PR myself, and the PR lost my attention. I also missed the notification after you deleted the `bin` directory.  Well, that’s a terrible situation. I regret I didn’t notice all of that earlier. I’m really, really sorry for this. Please accept my apologies, @tnsilver. Tradukisto is currently getting a new maintainer and my approval was meant to be an information for the future maintainer, and a sign that your PR, hopefully, is finally going to master soon.

MaciejLipinski commented:
Hello @tnsilver , I'm the new maintainer of tradukisto. I share @wjur 's regret about the delay. If you could reopen the pull request, I'll take care of it and merge it without further ado.

tnsilver commented:
Hi, it's been too long and I've changed my mind about this project. I no longer wish to take part in it. Best of luck in your new capacity.

MaciejLipinski commented:
> Hi, it's been too long and I've changed my mind about this project. I no longer wish to take part in it. Best of luck in your new capacity.

Would you mind if I copied your code to a branch on the source repository and merged it there? If you agree, would you like me to add your name in `// @author` comments (since git blame would not show your authorship)?

tnsilver commented:
I don't mind but I've deleted my git repo months ago, so unless you have
the source code - it's gone.
Regards, Tom.

On Tue, Oct 5, 2021 at 5:19 PM Maciej Lipiński ***@***.***>
wrote:

> Hi, it's been too long and I've changed my mind about this project. I no
> longer wish to take part in it. Best of luck in your new capacity.
>
> Would you mind if I copied your code to a branch on the source repository
> and merged it there? If you agree, would you like me to add your name in //
> @author comments (since git blame would not show your authorship)?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/allegro/tradukisto/pull/38#issuecomment-934457052>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAB2QHPRCTWCRZWUEKDM6HLUFMCOFANCNFSM4PYQ6XGA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>
